### PR TITLE
ElementDeletionGeneratorBase delete_exteriors option

### LINKED ISSUE #24611: Element deletion generators should be able to preserve boundary elements
## Reason
Currently, when requested to delete an element, an ElementDeletionGeneratorBase subclass will also delete any boundary elements which have the requested element as an `interior_parent()`.  This is desirable in some contexts but not as a general rule; being able to replace a mesh with its boundary mesh (by first generating the boundary elements then deleting the interior elements) is a useful feature in some workflows.

## Design
Add an option (defaulting to the current behavior) to disable this behavior.

## Impact
No impact to anyone who doesn't invoke the new option.

### PR DESCRIPTION
**Pull request by roystgnr**
Closes #24611

Pinging @shikhar413 - try using or locally merging this branch, adding `delete_exteriors = false` to your own interior element deletion generator, and see if that gets you a boundary suitable for XYDelaunay?

### CODE REVIEW

File: framework/src/meshgenerators/ElementDeletionGeneratorBase.C [hunk 1]
```diff
@@ -145,6 +155,10 @@ ElementDeletionGeneratorBase::generate()
       for (unsigned int n = 0; n != n_sides; ++n)
         if (elem->neighbor_ptr(n) == remote_elem)
           queries[pid].push_back(std::make_pair(elem->id(), n));
+
+      // Use an OOB side index to encode "interior_parent"
```
Comment (lindsayad) [inline framework/src/meshgenerators/ElementDeletionGeneratorBase.C:159]:
```suggestion
      // Use an OOB side index to encode "interior_parent". We will use this OOB index later
```

File: framework/src/meshgenerators/ElementDeletionGeneratorBase.C [hunk 2]
```diff
@@ -181,9 +195,10 @@ ElementDeletionGeneratorBase::generate()
       {
         const Elem * elem = mesh->elem_ptr(q.first);
         const unsigned int side = q.second;
-        const Elem * neighbor = elem->neighbor_ptr(side);
+        const Elem * target =
+            (side > elem->n_sides()) ? elem->interior_parent() : elem->neighbor_ptr(side);
```
Comment (lindsayad) [inline framework/src/meshgenerators/ElementDeletionGeneratorBase.C:199]:
```suggestion
            (side >= elem->n_sides()) ? elem->interior_parent() : elem->neighbor_ptr(side);
```

Comment (lindsayad) [reply framework/src/meshgenerators/ElementDeletionGeneratorBase.C:199]: I think this change is correct?

Comment (roystgnr) [reply framework/src/meshgenerators/ElementDeletionGeneratorBase.C:199]: What the hell.  It's correct, and it's proof that my test case isn't testing as much as I'd hoped.  I guess I'd need to specifically cook up something where we're ghosting a boundary element but not ghosting its `interior_parent`?  Not sure how to force that; I'd been hoping that the tests I ran on my own (throwing that unit test at 64 processors to chop it up as finely as I could) would do it.

Comment (lindsayad) [reply framework/src/meshgenerators/ElementDeletionGeneratorBase.C]: You could write a custom `RelationshipManager` that you put in the `test` directory that does that

**Comment by shikhar413** [comment]
@roystgnr still seeing an empty mesh upon deletion and setting `delete_exteriors` to false. Here is my input file:

```assembly_pitch = 32.353

[Mesh]
  [dummy]
    type = PolygonConcentricCircleMeshGenerator
    num_sides = 6
    num_sectors_per_side = '4 4 4 4 4 4'
    polygon_size = ${fparse assembly_pitch/2.0}
    polygon_size_style = 'apothem'
    background_block_ids = '99'
  []
  [inner_ref_2d]
    type = PatternedHexMeshGenerator
    inputs = 'dummy'
    pattern_boundary = none
    generate_core_metadata = false
    generate_control_drum_positions_file = false
   pattern =
'        0  0  0  0;
       0  0  0  0  0;
      0  0  0  0  0  0;
     0  0  0  0  0  0  0;
      0  0  0  0  0  0;
       0  0  0  0  0;
        0  0  0  0'
    external_boundary_name = "VOID2"
    rotate_angle = 0
  []
  [lower_d_inner]
    type = LowerDBlockFromSidesetGenerator
    input = inner_ref_2d
    new_block_name = 'inner_ref_interface'
    sidesets = 'VOID2'
    delete_exteriors = false
  []
  [del_lower_d_inner_block]
    type = BlockDeletionGenerator
    input = lower_d_inner
    block = 99
  []
  uniform_refine = 0
[]
```

**Comment by roystgnr** [comment]
Move `delete_exteriors = false` to the `BlockDeletionGenerator`?

**Comment by lindsayad** [comment]
Yea from the coverage report it looks like we are not hitting some of the important parallel changes

**Comment by shikhar413** [comment]
> 

Right, thought this was a LowerDBlockFromSidesetGenerator parameter. XYDelaunayGenerator works flawlessly now and I can generate the following meshes without any nonconformal nodes:




Thanks for the help!

**Comment by roystgnr** [comment]
> Yea from the coverage report it looks like we are not hitting some of the important parallel changes

Line 161 not covered.  MOOSE isn't ghosting all interior parents of ghosted elements, is it?

**Comment by lindsayad** [comment]
geometric ghosting should just be the default libMesh geometric ghosting functor

**Comment by roystgnr** [comment]
Oh, man.  That's enough, isn't it?  We ghost point neighbors, and unless someone's been going crazy overriding adaptive mesh refinement smoothing rules, every boundary element shares at least one of its points with its interior_parent().  And ... MOOSE doesn't even have an interface to `libMesh::MeshRefinement::overrefined_boundary_limit()`.  I don't think there's any way to get test coverage of that line without adding additional adaptivity parameters, which I'd rather not do just to get a completely unrelated feature working.  Should I leave that code as-is for attempted future-proofing, or take it out and replace it with some assertions that it's not needed?

**Comment by lindsayad** [comment]
I feel like leaving it in for future proofing is a good way to go

**Review by lindsayad — APPROVED** [review]

**Pull request merged**